deep_qa/layers/backend/expand_from_batch.py

```diff
@@ -0,0 +1,71 @@
+from overrides import overrides
+import tensorflow as tf
+
+from ..masked_layer import MaskedLayer
+
+
+class ExpandFromBatch(MaskedLayer):
+    """
+    Reshapes a collapsed tensor, taking the batch size and separating it into ``num_to_expand``
+    dimensions, following the shape of a second input tensor.  This is meant to be used in
+    conjunction with :class:`~deep_qa.layers.backend.collapse_to_batch.CollapseToBatch`, to achieve
+    the same effect as Keras' ``TimeDistributed`` layer, but for shapes that are not fully
+    specified at graph compilation time.
+
+    For example, say you had an original tensor of shape (2, 4, 5, 3), then collapsed it with
+    ``CollapseToBatch(2)(tensor)`` to get a tensor with shape (40, 3).  You can then call
+    ``ExpandFromBatch(2)(collapsed, tensor)`` with the result to expand the first two dimensions
+    out of the batch again (presumably after you've done some computation when it was collapsed).
+
+    Inputs:
+        - a tensor that has been collapsed with ``CollapseToBatch(num_to_expand)``.
+        - the original tensor that was used as input to ``CollapseToBatch`` (or one with identical
+          shape in the collapsed dimensions).  We will use this input only to get its shape.
```

Comment:
Because the shape will have `Nones` in it - that's the whole point of using this instead of `TimeDistributed`.  We need a way to get runtime shape, which we do by calling `tf.shape` on the input tensor.  We don't have a good way to get runtime shape in `_build_model`, where we're dealing with Keras layers, so we just take the original tensor to get the shape, so we can expand back the same way.  This should be a bit more clear once I add some comments to `CollapseToBatch`.